### DISCUSSION TITLE
Purpose of gitignore and interplay with git clean

### DISCUSSION THREAD

**Discussion by tiplu-alexpovel** [alirezanet/Husky.Net#59] [Q&A] [answered]
Hi!

Thanks you for creating this tool. We use it to keep our commits straight and catch build-breaking mistakes locally, before they even hit CI.

We're having a small issue regarding the `.husky/_` directory. It's `gitignored` by itself, hence a `git clean -xfd` will get rid of it. We use the latter regularly for all sorts of useful tasks. From your source code I can see you're using the `_` directory to hold a cache, among other things. It also holds `husky.sh`, which is vital for operation. Hence a `git clean -xfd` will always need to be followed by a `dotnet husky install` (the times I've forgotten this are numerous...). Otherwise, hooks will fail on the next git action triggering any:

```bash
$ git commit -am "test"
.husky/prepare-commit-msg: line 2: .husky/_/husky.sh: No such file or directory
```

Would you happen to have a solution for this? Can we perhaps safely un-ignore `.husky/_/husky.sh`? Its contents seem hard-coded anyway, unlike a cache or dynamic files each developer would have to generate for their local OS etc. (hence files that shouldn't be in VCS):

https://github.com/alirezanet/Husky.Net/blob/ca3543d482b6d6aa916116b348d22a74def9129a/src/Husky/Cli/InstallCommand.cs#L109-L116

**Comment by alirezanet** [comment, marked as answer]
Hi @tiplu-alexpovel,

These files are ignored by default because usually, people don't want to push the husky internal files to their repository. However, you have raised an interesting issue! Removing the .gitignore file might solve the problem, but it will probably be regenerated each time you use the husky install command (I haven't tested it yet). This isn't the best solution if it gets regenerated.

However, I can add an option to the install command to prevent generating the .gitignore file if you think it could solve the problem. Please let me know if you think this option could work, and if you have any other ideas or suggestions for resolving the issue. I will do my best to implement it as soon as possible.

**Reply by alirezanet** [reply]
PS. I'm currently on vacation and unfortunately, I don't have access to my computer to test it out. However, I will be back next week and will be able to check this out.

**Comment by tiplu-alexpovel** [comment]
Hi,

thanks for your super swift reply. Please enjoy your vacation, we're between years after all and this issue is not urgent!

You're right, a `dotnet husky install` will regenerate the `.gitignore` file. However, deleting it and committing `husky.sh` *once* still works. The file is henceforth part of the repo and even regenerating `.gitignore` and then *changing* `husky.sh` continues to work (new changes are picked up for staging & comitting). As such, even if you introduced upstream changes to that file, they could propagate to us (instead of being gitignored perpetually).

A flag to skip generating a `.gitignore` would therefore be obsolete, although cleaner.

For us, it's enough to know committing `husky.sh` isn't actively harmful 🙂
